### PR TITLE
Create user resolver

### PR DESCRIPTION
**Pull request by mashish584**
Create user resolver setup with email verification using sendGrid and validation check also placed.

### CODE REVIEW

File: server/package.json [hunk 1]
```diff
@@ -13,14 +13,19 @@
...
-		"graphql-yoga": "^1.18.3"
+		"graphql-yoga": "^1.18.3",
+		"uuidv4": "^6.2.7"
 	},
 	"devDependencies": {
 		"@graphql-cli/codegen": "^1.17.22",
 		"@graphql-codegen/schema-ast": "^1.18.1",
 		"@graphql-codegen/typescript-resolvers": "^1.19.0",
+		"@typescript-eslint/eslint-plugin": "^4.22.0",
+		"@typescript-eslint/parser": "^4.22.0",
```
Comment (ilakshay14) [inline server/package.json:28]: Add unit testing as well!

File: server/src/index.ts [hunk 2]
```diff
@@ -1,9 +1,20 @@
...
 
+const prisma = new PrismaClient();
+
+export type prismaClientType = typeof prisma;
+
 const server = new GraphQLServer({
   typeDefs: "./src/schema.graphql",
   resolvers,
+  context(request) {
+    return {
```
Comment (ilakshay14) [inline server/src/index.ts:13]: typescript doesn't support inline returns?

And if you want, you can add an api check here. For example, add an X API KEY. Use bcrypt to decrypt and check and if the secret API key matches, then only allow the request to go ahead, else throw an unauth error.

File: server/src/resolvers/mutation.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import { createUserSchema } from "../utils/validationSchemas";
+
+const Mutation = {
+  async createUser(parent, args, { prisma, request }: contextType, info): Promise<signupPayload> {
+    const validationSchema: Array<ValidationBody> = createUserSchema(args.data);
+
+    const { errors } = validate(validationSchema);
+
+    if (Object.keys(errors).length) {
+      throw new Error("Please validate all inputs.");
```
Comment (ilakshay14) [inline server/src/resolvers/mutation.ts:15]: GraphQL-YOGA doesnt provide any out of the box error formatters?

Comment (ilakshay14) [reply server/src/resolvers/mutation.ts:15]: if not, would suggest to create error helpers such as UserInput error or Auth error. Check Apollo server for the concept.

File: server/src/resolvers/mutation.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    });
+
+    if (isEmailExist) {
+      throw new Error("Email address is not available.");
+    }
+
+    try {
+      const { origin } = request.request.headers;
+      const verificationCode = uuid();
+      const password = bcrypt.hashSync(args.data.password, 10);
```
Comment (ilakshay14) [inline server/src/resolvers/mutation.ts:43]: I feel using await for bcrypt functions will also help. Not sure, check once.

File: server/src/resolvers/mutation.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      const user = await prisma.user.create({
+        data: {
+          verificationCode,
+          password,
+          ...args.data,
+        },
+      });
+
+      await sendMail({
+        from: "ep@mailinator.com",
```
Comment (ilakshay14) [inline server/src/resolvers/mutation.ts]: use env to store this email.

File: server/src/resolvers/query.ts [hunk 6]
```diff
@@ -1,5 +1,10 @@
+import { User } from "@prisma/client";
+import { contextType } from "../types";
+
 const Query = {
-  hello: (_, { name }) => `👋 ${name}`,
+  users(parent, args, { prisma }: contextType, info): Promise<User[]> {
```
Comment (ilakshay14) [inline server/src/resolvers/query.ts:5]: async the queries as well.

File: server/src/schema.graphql [hunk 7]
```diff
@@ -1,3 +1,30 @@
 type Query {
-	hello(name: String): String!
+  users: [User!]
+}
+
+type Mutation {
+  createUser(data: CreateUserInput): SignupPayload!
```
Comment (ilakshay14) [inline server/src/schema.graphql:6]: no error for using CreateUserInput before defining it?

File: server/src/utils/validate.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
+import { firstUC } from ".";
+import { ValidationBody } from "../types";
+
+function validateEmail(email) {
+  const re = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```
Comment (ilakshay14) [inline server/src/utils/validate.ts]: use better variable name  (xD)

File: server/src/utils/validate.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  return re.test(String(email).toLowerCase());
+}
+
+export default (body: Array<ValidationBody>): { errors: object; values?: object } => {
```
Comment (ilakshay14) [inline server/src/utils/validate.ts:9]: { errors: object, or ;

File: server/src/utils/validate.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+  const errors = {};
+  let values = {};
+
+  if (!body.length) return { errors };
```
Comment (ilakshay14) [inline server/src/utils/validate.ts]: if body is empty, it will return empty object for errors.

File: server/src/utils/validate.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+  if (body.length) {
```
Comment (ilakshay14) [inline server/src/utils/validate.ts:13]: do you need to put this condition here? It the above fails, following block will considered.

File: server/src/utils/validate.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    values = body.map(({
+      config, key, value, type,
+    }) => {
+      const fieldName = config.title || firstUC(key);
+      value = value?.trim();
+      switch (type) {
+        case "email":
+          if (config.required) {
+            if (value.length === 0) {
```
Comment (ilakshay14) [inline server/src/utils/validate.ts]: you can combine these two conditions

File: server/src/utils/validate.ts [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+              errors[key] = `${fieldName} is required.`;
+            }
+          }
+
+          if (!errors[key]) {
+            const isValid = validateEmail(value);
```
Comment (ilakshay14) [inline server/src/utils/validate.ts:26]: same, you can combine here as well

File: server/src/utils/validate.ts [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+            if (!isValid) {
+              errors[key] = "Please provide valid email address";
+            }
+          }
+          break;
+        case "password":
+        case "text":
```
Comment (ilakshay14) [inline server/src/utils/validate.ts:37]: case email can also be combined with this, unless you have something diff planned for email case

File: server/src/utils/validationSchemas.ts [hunk 15]
```diff
@@ -0,0 +1,39 @@
+import { config, ValidationBody, validationBodyType } from "../types";
+
+const getTypeOfField = (key): validationBodyType => {
+  const defaultField: Array<validationBodyType> = ["email", "password"];
+  if (defaultField.includes(key)) {
```
Comment (ilakshay14) [inline server/src/utils/validationSchemas.ts:5]: you can directly use the array instead of creating a variable

**Review by ilakshay14 — CHANGES_REQUESTED** [review]
Have added some comments. If they make sense, go for it.

File: server/package.json (hunk 1 above)
Comment (mashish584) [reply server/package.json:28]: I will add that once I will done with all modules.

File: server/src/utils/validate.ts (hunk 9 above)
Comment (ilakshay14) [reply server/src/utils/validate.ts:9]: shouldn't export default have a comma here? { errors: object, -> ye wala

**Review by ilakshay14 — COMMENTED** [review]
just verify that export default thing. Rest seems good to go.

**Review by ilakshay14 — APPROVED** [review]

**Pull request merged**